uint16 open_limit
uint16 close_limit